### PR TITLE
Better JSON Casting for Rows

### PR DESCRIPTION
**Pull request by pamitm-uber**
This commit is cherry-picked from https://github.com/trinodb/trino/pull/3613/files

In the existing code, rows are converted to JSON arrays. If the row has a lot of fields, then it is really hard to read the JSON output to understand which field each value in the array corresponds to.
Instead, this commit, casts rows as JSON maps with the field name as the key in the map and the field value as the value.

Test plan
1. Unit Tests
2. Ran Presto and queried an internal data source 

```
== RELEASE NOTES ==

General Changes
* Improved row to json cast by including field names in the resulting json by making it a map instead of an array.
```

### CODE REVIEW

**Comment by pamitm-uber** [comment]
Not sure why the presto-docs build is failing in CI. It passes locally.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java
Comment (pamitm-uber) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:499]: This is the part which changes the default to be the new json converted which makes it a breaking change. We can change the default to use legacy cast, if needed

**Comment by pamitm-uber** [comment]
@ajaygeorge is it possible to get this reviewed by someone else? After, I get the merge conflicts fixed of course.

**Comment by pamitm-uber** [comment]
@pranjalssh Could you please take a look at this?

**Comment by pamitm-uber** [comment]
@ajaygeorge Is there anyone else who can help us with this review? Pranjal/Sreeni seem to be very busy 😅

**Comment by beinan** [comment]
I rerun the failed test a couple of times, but it's still failing.  Probably a rebase is needed

**Comment by pamitm-uber** [comment]
@beinan I rebased again but there are a few tests which randomly fail every time.

**Comment by shangxinli** [comment]
Thanks @beinan! After some work, Amit is able to pass the tests. Can you have a look?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java
Comment (shangxinli) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java:80]: Do we check whether or not typeSignatureParameters has the same or large size of fieldTypes.size()? Otherwise could we get out of range?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java
Comment (shangxinli) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java:79]: Can we do something like below? 

List<String> fieldNames = IntStream.range(0, fieldTypes.size())
    .mapToObj(i -> typeSignatureParameters.get(i).getNamedTypeSignature().getName().orElse(""))
    .collect(Collectors.toList());

IntStream.range(0, fieldTypes.size())
    .forEach(i -> fieldWriters.add(createJsonGeneratorWriter(fieldTypes.get(i))));

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java
Comment (shangxinli) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java:122]: Need some explanation of the '40'.

Comment (pamitm-uber) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java:122]: I cannot find any reasoning for why 40 was chosen initially but my guess is that this is a good proxy for the average size of the field value. Given that we will now be adding the fieldname to the output as well, I think I should make the estimated size of the slice 50(40 chars for value + 10 chars for the key). What do you think?

 
PR which introduced the change for context: https://github.com/prestodb/presto/pull/7565/files

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java
Comment (pamitm-uber) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java:80]: For rows, typeSignature and typeParameters are both transformation of the fields so they will have the same size

Please see: https://github.com/prestodb/presto/blob/master/presto-common/src/main/java/com/facebook/presto/common/type/RowType.java

Comment (shangxinli) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java]: Do we want to add assert that the two size are same?

Comment (pamitm-uber) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToJsonCast.java]: Done

**Comment by shangxinli** [comment]
LGTM

File: presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java
Comment (beinan) [inline presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java:666]: will the default value "" cause trouble such as duplicated field name?

**Review by beinan — COMMENTED** [review]
Great contribute, just a couple of style issue and clarification qustion.

File: presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java
Comment (pamitm-uber) [reply presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java:666]: Nope, we already have a  test for that in https://github.com/prestodb/presto/pull/19513/files#diff-9cb149b592e20c436e946cf96f593ae4f1fbf948513df713f71bdcd772279324R154

File: presto-common/src/main/java/com/facebook/presto/common/function/SqlFunctionProperties.java
Comment (pamitm-uber) [reply presto-common/src/main/java/com/facebook/presto/common/function/SqlFunctionProperties.java]: My bad removed, old habits :)

**Review by beinan — APPROVED** [review]
lgtm

File: presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java
Comment (beinan) [reply presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java:666]: Great, then I have no concern

**Review by beinan — CHANGES_REQUESTED** [review]
looks all good, but could you squash and push? thanks!

**Review by beinan — APPROVED** [review]
awesome!! thanks, I will merge it once the CI finish

**Pull request merged**
